nexus/src/db/datastore/silo_user.rs

```diff
@@ -202,6 +212,96 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
+    /// Updates or deletes the password hash for a given Silo user
+    ///
+    /// If `password_hash` is `Some(...)`, the provided value is stored as the
+    /// user's password hash.  Otherwise, any existing hash is deleted so that
+    /// it cannot be used for authentication any more.
```

Comment:
Yeah, fair question.  I don't think it'd make things simpler or safer.  I think it'd be confusing for the database state to suggest that there was a password set when none had ever been set.  In practice, I also don't like this because I think it would make this subsystem harder to debug and support if we cannot tell whether a password is set.  One could argue that it shouldn't matter -- that if somebody raises a ticket because they cannot authenticate, the answer should always be "verify them some other way and reset the password".  But we may have legitimate bugs where people fail to authenticate and it'll be harder to root-cause them if we can't tell which case we're in.  I could also imagine flows where we'd rather tell an operator "this user doesn't have a password set" because as an operator, I'd want some explanation for why something that I expect to work doesn't work rather than just being told to try again.

Reply:
Cool, thanks for thinking about it!